src/libclipper/include/clipper/task_executor.hpp

```diff
@@ -277,24 +324,38 @@ class TaskExecutor {
   redox::Redox redis_connection_;
   redox::Subscriber redis_subscriber_;
   std::mutex inflight_messages_mutex_;
+  // inflight_messages_ value is vector of <message_id, model, input>
   std::unordered_map<int, std::vector<std::tuple<const long, VersionedModelId,
                                                  std::shared_ptr<Input>>>>
       inflight_messages_;
   std::shared_ptr<metrics::Counter> predictions_counter;
-  std::shared_ptr<metrics::Meter> throughput_meter;
-  std::shared_ptr<metrics::Histogram> latency_hist;
+  std::shared_ptr<metrics::Meter> throughput_meter_;
+  boost::shared_mutex model_queues_mutex_;
   std::unordered_map<const VersionedModelId, ModelQueue,
                      decltype(&versioned_model_hash)>
       model_queues_;
+  boost::shared_mutex model_metrics_mutex_;
+  std::unordered_map<const VersionedModelId, ModelMetrics,
+                     decltype(&versioned_model_hash)>
+      model_metrics_;
   static constexpr int INITIAL_MODEL_QUEUES_MAP_SIZE = 100;
 
   bool create_model_queue_if_necessary(const VersionedModelId &model_id) {
     // Adds a new <model_id, task_queue> entry to the queues map, if one
     // does not already exist
+    boost::unique_lock<boost::shared_mutex> l(model_queues_mutex_);
     auto queue_added = model_queues_.emplace(std::piecewise_construct,
                                              std::forward_as_tuple(model_id),
                                              std::forward_as_tuple());
-    return queue_added.second;
+    bool queue_created = queue_added.second;
```

Comment:
It's worth noting that calls to `register_external_model` in `clipper_manager` won't result in the addition of new model metrics until a container associated with the given model actually connects to Clipper. Do you think this will cause user confusion ("Why aren't there any metrics for my model?")? 

Imo this behavior is fine.